### ISSUE TITLE
Execution of SPARQL query (as part of a reasoning task) does not finish on version 3.1.3 (rdf4j)

### ISSUE THREAD

**Issue by swc-kdzekov** [eclipse/rdf4j#2140]
Hello,
We updated our system rdf4j library from version 2.3.3 to version 3.1.3. Now, we are facing an issue running this query (which is part of reasoning process)

```
PREFIX h:<urn:pp:internalHelper/>
PREFIX owl:<http://www.w3.org/2002/07/owl#>
PREFIX rdf:<http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos:<http://www.w3.org/2004/02/skos/core#>
PREFIX skosxl:<http://www.w3.org/2008/05/skos-xl#>
PREFIX swcs:<http://schema.semantic-web.at/ppt/>
CONSTRUCT {?c a ?t. ?c a ?t2. } 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus> 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus/reasoning#7a9e2803-4ad7-4346-a9aa-e9fb4192b71b> 
WHERE {
    ?c a ?t.
    FILTER (?t IN (<https://pp-remotestore-test.semantic-web.at/SQA-GPO/GenericClass>))
    FILTER NOT EXISTS {
        ?c (swcs:appliedType|swcs:propagateType)/rdfs:subClassOf* ?t
    }
    FILTER NOT EXISTS {
        ?c ^skos:member ?root. ?root swcs:appliedType/rdfs:subClassOf* ?t
    }
    FILTER NOT EXISTS {
        ?root a skos:OrderedCollection; 
                skos:memberList ?member; 
                <http://schema.semantic-web.at/ppt/appliedType>/rdfs:subClassOf* ?t. 
        ?member rdf:rest*/rdf:first ?c.     }
    FILTER NOT EXISTS {
        ?c (skos:broader|^skos:narrower)+ ?root. ?root swcs:propagateType/rdfs:subClassOf* ?t
    }
    FILTER NOT EXISTS {
        ?c (skos:broader|^skos:narrower)*/(skos:topConceptOf|^skos:hasTopConcept) ?root. ?root swcs:appliedType/rdfs:subClassOf* ?t
    }
    OPTIONAL {?t rdfs:subClassOf* ?t2.} 
}
```

The execution of the task does not end on the newer version (3.1.3) but it runs (it takes about 40 seconds) on the older Rdf4j version (2.3.3). Could you maybe give us some insights about what changed between the two versions that might affect the execution of the query?
P.S I have attached the data
[pp_project_mesh2016_sqagpsabclocalmemory (1).zip](https://github.com/eclipse/rdf4j/files/4539473/pp_project_mesh2016_sqagpsabclocalmemory.1.zip)

**Comment by hmottestad** [comment]
Rewriting all your FILTER NOT EXISTS together and changing your initial filter into a bind seems to make it fast:

```
PREFIX h:<urn:pp:internalHelper/>
PREFIX owl:<http://www.w3.org/2002/07/owl#>
PREFIX rdf:<http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos:<http://www.w3.org/2004/02/skos/core#>
PREFIX skosxl:<http://www.w3.org/2008/05/skos-xl#>
PREFIX swcs:<http://schema.semantic-web.at/ppt/>
CONSTRUCT {?c a ?t. ?c a ?t2. } 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus> 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus/reasoning#7a9e2803-4ad7-4346-a9aa-e9fb4192b71b> 
WHERE {
    BIND (<https://pp-remotestore-test.semantic-web.at/SQA-GPO/GenericClass> as ?t)
    ?c a ?t.
    FILTER NOT EXISTS {
        ?c (swcs:appliedType|swcs:propagateType)/rdfs:subClassOf* ?t .
        ?c ^skos:member ?root. ?root swcs:appliedType/rdfs:subClassOf* ?t .
        ?root a skos:OrderedCollection; 
                skos:memberList ?member; 
                <http://schema.semantic-web.at/ppt/appliedType>/rdfs:subClassOf* ?t. 
        ?member rdf:rest*/rdf:first ?c.     
        ?c (skos:broader|^skos:narrower)+ ?root. ?root swcs:propagateType/rdfs:subClassOf* ?t.
        ?c (skos:broader|^skos:narrower)*/(skos:topConceptOf|^skos:hasTopConcept) ?root. ?root swcs:appliedType/rdfs:subClassOf* ?t.
    }
    OPTIONAL {?t rdfs:subClassOf* ?t2.} 
}
```

Does this return the same results or am I missing something?

**Comment by swc-kdzekov** [comment]
Hi,
Thanks for the feedback. Truly, it runs fast but returns different results.
I'll try to fix the query using your approach.

**Comment by hmottestad** [comment]
@jeenbroekstra I am wondering if there is an edge case in your code from #1405 that isn't handled correctly.

The original query has no unions in it. But the query plan introduces unions. The code in `isOutOfScopeForLeftArgBindings` in JoinIterator sees this union and uses the MergeIteration (which is actually a loop join I think). 

When I remove `isOutOfScopeForLeftArgBindings` the query runs at the same speed as it used to.

I've created a simplified version of this query that exhibits the same issue:

```
PREFIX h:<urn:pp:internalHelper/>
PREFIX owl:<http://www.w3.org/2002/07/owl#>
PREFIX rdf:<http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos:<http://www.w3.org/2004/02/skos/core#>
PREFIX skosxl:<http://www.w3.org/2008/05/skos-xl#>
PREFIX swcs:<http://schema.semantic-web.at/ppt/>
CONSTRUCT {?c a ?t. ?c a ?t2. } 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus> 
FROM <https://performance.sqa.semantic-web.at/MeSH2016-SQAGPS_ABC-localMemory/thesaurus/reasoning#7a9e2803-4ad7-4346-a9aa-e9fb4192b71b> 
WHERE {
    BIND (<https://pp-remotestore-test.semantic-web.at/SQA-GPO/GenericClass> as ?t)
    ?c a ?t.
    FILTER NOT EXISTS {
        ?c (skos:broader|^skos:narrower)*/(skos:topConceptOf|^skos:hasTopConcept) ?root. ?root swcs:appliedType/rdfs:subClassOf* ?t
    }
}
```

### Query plan
```
Reduced
   MultiProjection
      ProjectionElemList
         ProjectionElem "c" AS "subject"
         ProjectionElem "_const_f5e5585a_uri" AS "predicate"
         ProjectionElem "t" AS "object"
      ProjectionElemList
         ProjectionElem "c" AS "subject"
         ProjectionElem "_const_f5e5585a_uri" AS "predicate"
         ProjectionElem "t2" AS "object"
      Extension
         ExtensionElem (_const_f5e5585a_uri)
            ValueConstant (value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type)
         ExtensionElem (t2)
            Var (name=t2)
         Filter
            Not
               Exists
                  Join
                     StatementPattern (costEstimate=1, resultSizeEstimate=3)
                        Var (name=root)
                        Var (name=_const_901ac966_uri, value=http://schema.semantic-web.at/ppt/appliedType, anonymous)
                        Var (name=_anon_af88f1a0_d7de_40f1_802a_01979e3769d7, anonymous)
                     Union <-- This here is the first union that becomes a problem.
                        Join
                           StatementPattern (costEstimate=2, resultSizeEstimate=52)
                              Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                              Var (name=_const_e8cbb1af_uri, value=http://www.w3.org/2004/02/skos/core#topConceptOf, anonymous)
                              Var (name=root)
                           Join
                              ArbitraryLengthPath (costEstimate=710, resultSizeEstimate=504.3K)
                                 Var (name=c)
                                 Union
                                    StatementPattern (resultSizeEstimate=13.5K)
                                       Var (name=c)
                                       Var (name=_const_7123f66a_uri, value=http://www.w3.org/2004/02/skos/core#broader, anonymous)
                                       Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                                    StatementPattern (resultSizeEstimate=13.5K)
                                       Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                                       Var (name=_const_14b25ded_uri, value=http://www.w3.org/2004/02/skos/core#narrower, anonymous)
                                       Var (name=c)
                                 Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                              ArbitraryLengthPath (costEstimate=710, resultSizeEstimate=504.3K)
                                 Var (name=_anon_af88f1a0_d7de_40f1_802a_01979e3769d7, anonymous)
                                 StatementPattern (resultSizeEstimate=0)
                                    Var (name=_anon_af88f1a0_d7de_40f1_802a_01979e3769d7, anonymous)
                                    Var (name=_const_4592be07_uri, value=http://www.w3.org/2000/01/rdf-schema#subClassOf, anonymous)
                                    Var (name=t)
                                 Var (name=t)
                        Join
                           StatementPattern (costEstimate=2, resultSizeEstimate=52)
                              Var (name=root)
                              Var (name=_const_d2459908_uri, value=http://www.w3.org/2004/02/skos/core#hasTopConcept, anonymous)
                              Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                           Join
                              ArbitraryLengthPath (costEstimate=710, resultSizeEstimate=504.3K)
                                 Var (name=c)
                                 Union
                                    StatementPattern (resultSizeEstimate=13.5K)
                                       Var (name=c)
                                       Var (name=_const_7123f66a_uri, value=http://www.w3.org/2004/02/skos/core#broader, anonymous)
                                       Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                                    StatementPattern (resultSizeEstimate=13.5K)
                                       Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                                       Var (name=_const_14b25ded_uri, value=http://www.w3.org/2004/02/skos/core#narrower, anonymous)
                                       Var (name=c)
                                 Var (name=_anon_7abeb157_cd35_46c5_b76d_5332385f2f03, anonymous)
                              ArbitraryLengthPath (costEstimate=710, resultSizeEstimate=504.3K)
                                 Var (name=_anon_af88f1a0_d7de_40f1_802a_01979e3769d7, anonymous)
                                 StatementPattern (resultSizeEstimate=0)
                                    Var (name=_anon_af88f1a0_d7de_40f1_802a_01979e3769d7, anonymous)
                                    Var (name=_const_4592be07_uri, value=http://www.w3.org/2000/01/rdf-schema#subClassOf, anonymous)
                                    Var (name=t)
                                 Var (name=t)
            Join
               Extension
                  ExtensionElem (t)
                     ValueConstant (value=https://pp-remotestore-test.semantic-web.at/SQA-GPO/GenericClass)
                  SingletonSet
               StatementPattern (costEstimate=20.5K, resultSizeEstimate=20.5K)
                  Var (name=c)
                  Var (name=_const_f5e5585a_uri, value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type, anonymous)
                  Var (name=t)
```

**Comment by jeenbroekstra** [comment]
You're correct that path alternatives are internally treated as unions.

I don't think it's so much an edge case that we overlooked, but more that #1405 _itself_ is an edge case. The fix we introduced to make sure we are correct according to the spec is really only relevant in very specific corner cases, and unfortunately it means evaluation of unions generally is a lot less efficient (as this shows). 

Currently, we decide to use a merge join iteration based purely on the fact that something is a union. I'm wondering if we can do slightly smarter evaluation by inspecting the contents of the union clauses. I haven't thought this through fully but I wonder if we really only need to revert to merging afterwards if the union contains BIND operations in one of its arguments, or if it filters on variables that are not (also) bound inside the union itself.

**Comment by jeenbroekstra** [comment]
I'm also looking at why we have this logic in the JoinIterator itself in the first place, instead of in StrictEvaluationStrategy, and why/how this case is different from the decision to use a HashJoinIterator instead of a normal JoinIterator (which we do for subqueries).

**Comment by hmottestad** [comment]
I was thinking of maybe adding a “new scope” flag to union. Default to true. But if we introduce a union because of a path query we can set it to false. We could then introduce an optimizer that could optimize other unions by checking their sub tree for potential scoping issues.

**Comment by jeenbroekstra** [comment]
We can use the existing is `isGraphPatternGroup` property for that I think.

**Comment by swc-kdzekov** [comment]
Hi, just curious to know, is the bug going to be fixed for the next version (3.1.5) ?
Put it in the other way, could we expect that in the next version of rdf4j we would be able to run the query posted ?

**Comment by hmottestad** [comment]
Yes. @jeenbroekstra fixed the underlying bug so it should have the same performance. Would you test that it works for you by doing:

```
git clone https://github.com/eclipse/rdf4j.git
mvn clean install -DskipTests
```

Then use this tag in your project: `3.1.5-SNAPSHOT`

**Comment by swc-kdzekov** [comment]
Pulled,  I clean/install, but i don't know what jar file to include in my tomcat library. Before we used "eclipse-rdf4j-3.1.3-onejar.jar"

**Comment by swc-kdzekov** [comment]
I read in the documentation. I will run: "mvn -Passembly package", and look for onejar file.

**Comment by swc-kdzekov** [comment]
Yes, i pulled and clean/install it, but still, my project requires onejar file. And i can not build it currently using the command "mvn -Passembly package". It throws errors.

**Comment by hmottestad** [comment]
I'll try to create one for you

**Comment by hmottestad** [comment]
[eclipse-rdf4j-3.2.0-SNAPSHOT-onejar.jar.zip](https://github.com/eclipse/rdf4j/files/4564103/eclipse-rdf4j-3.2.0-SNAPSHOT-onejar.jar.zip)

**Comment by swc-kdzekov** [comment]
Thank you very much

**Comment by swc-kdzekov** [comment]
Hi again,
Using the git repository you shared i built the necessary artifacts and updated the system to 3.1.5-SNAPSHOT.
I **confirm** that the posted query executes, that is the good part.
However, the execution time comparing to an older version (2.3.3) is much longer.
For the same data corpus, on version 3.1.5-SNAPSHOT takes above 3 minutes, while on version 2.3.3 it takes 46 seconds.

**Comment by hmottestad** [comment]
There will be some documentation with the 3.2.0 release on how to use the new query .explain() feature. Hopefully that will make diagnosing this issue easier. We are also working on two improvements, one to the way we handle scoping or variables when parsing the SPARQL queries and another that exposes the this scoping in the query plan.

**Comment by swc-kdzekov** [comment]
Alright, thank you. I tested it again on Linux server, and the query execution was even faster. 
I have one more question regarding the future releases, if you please.

Is there going to be official release for 3.1.5 version ?

Knowing this we would be able to have clear picture on how we will continue. 
Thank you in advance.

**Comment by hmottestad** [comment]
All the fixes are included in the new 3.2.0 release.
